ompi/request/grequest.c

```diff
@@ -24,33 +24,55 @@
 #include "ompi/request/grequest.h"
 #include "ompi/mpi/fortran/base/fint_2_int.h"
 
-/**
+/*
  * Internal function to specialize the call to the user provided free_fn
  * for generalized requests.
  * @return The return value of the user specified callback or MPI_SUCCESS.
  */
-static inline int ompi_grequest_internal_free(ompi_grequest_t* greq)
+static inline int ompi_grequest_invoke_free(ompi_grequest_t* greq)
 {
-    int rc = MPI_SUCCESS;
+    int rc = OMPI_SUCCESS;
+    MPI_Fint ierr;
+
     if (NULL != greq->greq_free.c_free) {
+        if (greq->greq_funcs_are_c) {
+            rc = greq->greq_free.c_free(greq->greq_state);
+        } else {
+            greq->greq_free.f_free((MPI_Aint*)greq->greq_state, &ierr);
+            rc = OMPI_FINT_2_INT(ierr);
+        }
         /* We were already putting query_fn()'s return value into
          * status.MPI_ERROR but for MPI_{Wait,Test}*.  If there's a
          * free callback to invoke, the standard says to use the
          * return value from free_fn() callback, too.
          */
-        if (greq->greq_funcs_are_c) {
-            greq->greq_base.req_status.MPI_ERROR =
-                greq->greq_free.c_free(greq->greq_state);
-        } else {
-            MPI_Fint ierr;
-            greq->greq_free.f_free((MPI_Aint*)greq->greq_state, &ierr);
-            greq->greq_base.req_status.MPI_ERROR = OMPI_FINT_2_INT(ierr);
+        if (OMPI_SUCCESS != rc) {
+            greq->greq_base.req_status.MPI_ERROR = rc;
         }
-        rc = greq->greq_base.req_status.MPI_ERROR;
     }
     return rc;
  }
 
+
+/*
+ * Internal function to dispatch the call to the user provided free_fn
+ * for generalized requests. The freeing code executes as soon as both
+ * wait/free and complete have occured.
+ * @return The return value of the user specified callback or MPI_SUCCESS.
+ */
+static inline int ompi_grequest_internal_free(ompi_grequest_t* greq)
+{
+    int rc = OMPI_SUCCESS;
+
+    if (REQUEST_COMPLETE(&greq->greq_base) && greq->greq_user_freed) {
```

Comment:
Why do we only invoke free if the user called `MPI_Request_free` (i.e., from `ompi_grequest_free`)? Shouldn't free be invoked on completion (i.e., from `ompi_grequest_complete`) as well?

Reply:
@federeghe Please read the large comment below this function, startin with _MPI has some weird semantics with respect to generalized requests ..._.  And no, completion should not trigger free by itself, completion is sort of flagging the thing as done, such that the progress engine running in another thread knows that the thing is completed. What makes things convoluted is that a user could `Free` without a more usual `Wait` or `Test` call, therefore extra care is needed.